### ISSUE TITLE
Cannot find module 'pinkySwear' from '/home/message/www/jsapp/node_modules/qwest/build' 

### ISSUE THREAD

**Issue by message** [pyrsmk/qwest#71]
```
"qwest": "2.0.3"
```
```
[12:44 @ ~/www/jsapp] $ npm install
[12:44 @ ~/www/jsapp] $ gulp
[12:44:58] Using gulpfile ~/www/jsapp/gulpfile.js
[12:44:58] Starting 'lint'...
[12:44:58] Starting 'linttest'...
[12:44:58] Starting 'test'...
[12:44:58] Starting 'browserify'...
[12:44:58] Starting 'fonts'...
[12:44:58] Starting 'sprites'...
[12:44:58] Starting 'html'...
[12:44:58] Starting 'watch'...
[12:44:58] Finished 'watch' after 3.74 ms
[12:44:58] Starting 'server'...
[12:44:58] Finished 'server' after 6.2 ms
[12:44:58] Server started http://localhost:8000
[12:44:58] LiveReload started on port 35929
[12:44:59] Finished 'html' after 307 ms
debugger listening on port 5858
[12:45:00] Finished 'linttest' after 1.7 s
[12:45:00] Finished 'test' after 1.7 s
[12:45:01] Finished 'sprites' after 2.71 s
[12:45:01] Starting 'styles'...
[12:45:01] Finished 'fonts' after 2.84 s
[12:45:01] Finished 'styles' after 448 ms
[12:45:01] Starting 'images'...
[12:45:05] Finished 'images' after 3.32 s
[12:45:05] Finished 'lint' after 7.03 s
[12:45:06] [Error]: Cannot find module 'pinkySwear' from '/home/message/www/jsapp/node_modules/qwest/build' (undefined)
[12:45:06] Finished 'browserify' after 7.36 s
[12:45:06] Starting 'build'...
[12:45:06] Finished 'build' after 690 μs
[12:45:06] Starting 'default'...
[12:45:06] Finished 'default' after 5.03 μs
```

**Comment by pyrsmk** [comment]
PinkySwear and jquery-param are installed as dependencies in `qwest/node_modules/`. They are used to compile the minified version of qwest, available for browsers, and another version in `qwest/build/` for NPM.

Maybe your error is triggered by browserify because it is not configured as expected to handle qwest's dependencies. I have tested the NPM version of qwest in a project with browserify and all is working as expected.

**Comment by message** [comment]
how can i check that configuration? :S

**Comment by pyrsmk** [comment]
I don't know. It depends on how browserify is called (I have take a lot of time to have fully working browserify for my projects). Can you paste your browserify section?

**Comment by message** [comment]
```js
var _ = require("lodash");
var gulp = require("gulp");
var gutil = require("gulp-util");
var config = require("../config");
var watchify = require("watchify");
var uglify = require("gulp-uglify");
var connect = require("gulp-connect");
var browserify = require("browserify");
var notifier = require("node-notifier");
var streamify = require("gulp-streamify");
var source = require("vinyl-source-stream");
var messageformatify = require("../utils/messageformatify");

var isDevelopmentEnvironment = process.env.NODE_ENV === "development";
var sourceJs = [config.source, config.folders.javascripts, config.browserify.source].join("/");
var bundler = browserify(sourceJs, isDevelopmentEnvironment ? watchify.args : {});

if (isDevelopmentEnvironment) {
	bundler = watchify(bundler);
}

bundler.transform("reactify");
bundler.transform(messageformatify);
bundler.transform("brfs");

function bundle() {
	var destination = gulp.dest([config.build, config.folders.javascripts].join("/"));
	var bundle = bundler.bundle()
		.on("error", handleErrors)
		.pipe(source(config.browserify.source));

	// Uglify in production
	if (!isDevelopmentEnvironment) {
		bundle.pipe(streamify(uglify({
			// Uncomment next lines if you want beautified output
			//mangle: false,
			//output: {
			//	beautify: true
			//},
			compress: {
				drop_debugger: true,
				drop_console: true // Removes all console.* calls from code
			}
		}).on("error", handleErrors)));
	}

	// Write on disk
	bundle.pipe(destination);

	// Trigger reload in development
	if (isDevelopmentEnvironment) {
		bundle.pipe(connect.reload());
	}

	return bundle;
}

gulp.task("browserify", bundle); // Run task

function handleErrors(error) {
	notifier.notify({
		title: "Browsery error",
		message: "Error occured in broserify task. Check console."
	});

	// Notify console
	gutil.log(gutil.colors.red("[" + error.name + "]") + gutil.colors.white(": " + error.message + " (" + error.fileName + ")"));
	gutil.log(error);

	this.emit("end"); // Keep gulp from hanging on this task

	if (!isDevelopmentEnvironment) process.exit(1); // Exit with error when in production
}
```

**Comment by message** [comment]
Not sure how browserify is an issue here

**Comment by pyrsmk** [comment]
I don't know :s I don't use browserify like this.

```js
gulp.task('js', ['lint-js'], function() {

	var b,
		components = {},
		imports = {},
		streams = merge(),
		manifest,
		modules = {},
        name;

	// Generate module list
	require('./config/load.json').forEach(function(framework) {
        try {
            manifest = require('./'+framework+'/package.json');
        } catch (e) {}
        (_.keys(manifest.devDependencies) || []).forEach(function(name) {
            modules[name] = resolve.sync(name, {moduleDirectory: './'+framework+'/node_modules'});
        });
    });

	// Bundle vendor libraries
	b = browserify();
	b.add( './@source/various/ender.js' );
	for(name in modules){
		b.add(modules[name]);
		b.require(modules[name], {expose: name})
	}
	fs.closeSync( fs.openSync( './js/vendor.js', 'w' ) );
	streams.add(
		gulp.src( './js/vendor.js' )
			.pipe( plumber() )
			.pipe( transform(function(){ 
				return b.bundle();
			}) )
			.pipe( $if(prod, uglify()) )
			.pipe( gulp.dest('./js') )
			.pipe( size({gzip:true}) )
	);

	// Generate component list
	require('./config/load.json').forEach(function(framework) {
        fs.readdirSync('./'+framework+'/components').forEach(function(component) {
			fs.readdirSync('./'+framework+'/components/'+component).forEach(function(file) {
				if(/\.js$/.test(file)) {
					imports[framework+'/components/'+component+'/'+file] = './'+framework+'/components/'+component+'/'+file;
					if(component.charAt(0) == '@') {
						components[framework+'/'+component+'/'+file] = './'+framework+'/components/'+component+'/'+file;
					}
				}
			});
        });
    });

	// Bundle components
	for(name in components) {
        streams.add(
            gulp.src( components[name] )
                .pipe( plumber() )
                .pipe( transform(function(file) {
                    var name;
                    b = browserify(file);
                    for(name in modules) {
                        b.external(name);
                    }
					for(name in imports) {
						b.require(imports[name], {expose: name});
					}
                    for(name in components) {
                        b.require(components[name]);
                    }
                    return b.bundle();
                } ) )
                .pipe( $if(prod, uglify()) )
                .pipe( rename(name) )
                .pipe( gulp.dest('./js') )
                .pipe( size({gzip:true}) )
        );
	}

	return streams;

});
```

There's many components in my project, so I'm creating a list of all modules in the package.json of each framework to load (they are like components repository in fact). Then I bundle my vendor libraries, and then all the components in all loaded frameworks.

Anyway, as you see, I'm using `resolve` to get the path of each library in `package.json`. I'm also using `add()`, `require()` and `external()` (take a look at the browserify docs) to register my components exactly as they should be.

I don't know if it helps ^^

**Comment by pyrsmk** [comment]
Maybe the qwest's dependencies should be in the `build/` directory too...

**Comment by message** [comment]
the issue is with 2.0.3. 2.0.2 builds fine.

**Comment by pyrsmk** [comment]
2.0.2 is fine because the dependencies were directly concatenated to the library but they couldn't load in the browser. See #70

**Comment by message** [comment]
https://gist.github.com/message/60c5851f110dbcc0894a#file-qwest-min-js-L109 should be `pinkyswear`

**Comment by pyrsmk** [comment]
Nope. Should really be `require("pinkySwear")`, otherwise the library won't be found. Since the library is bundled by browserify, all libraries are wrapped and `window` is not accessible.

**Comment by message** [comment]
Well, i'm not so fluent in js, so i don't know, but library is not working for me with that change :S Let's see if someone will have same issue as me. I will stick with 2.0.2 for now

**Comment by message** [comment]
> otherwise the library won't be found
by whom? 

library saved as `pinkyswear` lowercase by node, there is no such library as pinkySwear...

**Comment by message** [comment]
Yeah, i get in browser with 2.0.2 `Uncaught ReferenceError: pinkySwear is not defined` and gulp is fine. With 2.0.3 changing `require("pinkySwear")` to `require("pinkyswear")` everything works fine :/

**Comment by message** [comment]
Ugh, somehow my `emit change` events not triggering. Will rollback to 1.7.0

**Comment by pyrsmk** [comment]
Ah, I completely misunderstood your correction. You're right, it's mispelled, I'm updating that now.

**Comment by pyrsmk** [comment]
I have published a new version on NPM, can you tell me if it's working now?

**Comment by pyrsmk** [comment]
Can you test on your side if you have time, so I can push the new version to github? :p

**Comment by message** [comment]
i had an issue with pinkyswear or something. My FLUX dispatcher wasn't dispatching any events, so i forked "fetchival" , made it look like qwest and added jquery-params to make a request data like i wanted in 1.7.0.

**Comment by pyrsmk** [comment]
Ah? Weird. What is FLUX?

**Comment by singerxt** [comment]
FLUX is Application Architecture https://facebook.github.io/react/docs/flux-overview.html

**Comment by pyrsmk** [comment]
Ah ok, thanks ^^

**Comment by pyrsmk** [comment]
So, I'm closing this ticket since it should be fixed with the last 2.0.4 release. Reopen it if needed ;)
